### ISSUE TITLE
[feature request] PyTorch support for sub-interpreters with PEP 684 accepted and release in Python 3.12

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#102517]
### 🚀 The feature, motivation and pitch

It appears that sub-interpretor support for Python in PEP 684 https://discuss.python.org/t/pep-684-a-per-interpreter-gil/ is advancing and will be merged soon (https://github.com/python/cpython/blob/main/Misc/NEWS.d/3.12.0b1.rst, https://github.com/python/cpython/pull/104210, https://github.com/ericsnowcurrently/multi-core-python/wiki/0-The-Plan)

Any updates on eventual support of sub-interpreters in PyTorch? 
I found older issues asking the same:
- https://github.com/pytorch/pytorch/issues/23490
- https://github.com/pytorch/pytorch/issues/10950

This might be important for working in hosted multi-threaded environments where every thread runs its own Python / PyTorch code. This is also related to the current torch.compile / export / mini-interpreters story: https://discuss.pytorch.org/t/torch-compiles-deployment-story-to-non-python-host-processes/180943. This scenario was explicitly supported with TorchScript / TorchJit, but seems not yet developed for torch.compile, but it's an important usecase too

**Comment by vadimkantorov** [comment]
@albanD I guess `multipy` would work! It would be great to see some tutorial of using `multiply` together with torch.compile / triton-compiled AOT packages

But maybe in the future, it's still important to track state-of-subinterpreters support for PyTorch/torch.compile and have a simple test

**Comment by vadimkantorov** [comment]
@albanD I would say this issue is not about python frontend, but more about "python backend" :)

**Comment by albanD** [comment]
>  It would be great to see some tutorial of using multipy together with torch.compile / triton-compiled AOT packages

They are independent so should "just work" in theory :)

**Comment by vadimkantorov** [comment]
Would be good to have a sample showcasing the two :) I think `multipy` is relatively unknown. And torch.compile currently misses communication about inference time (which was different for torch::jit::load etc)

**Comment by vadimkantorov** [comment]
I guess Python 3.12 is released now. So at least it would be nice to have some note in the docs on the support / interaction with sub-interpreters (even if it's not supported yet, it would be better to state it explicitly to save time for the users)

**Comment by albanD** [comment]
Note that we don't support 3.12 yet. But if you have a good grasp on sub-interpreters, feel free to write a blurb on it!

**Comment by vadimkantorov** [comment]
Regarding "better off using multiprocessing", this scenario is mainly about embedding python in some other inference process which might need multiple sub-interpreters (and then maybe spawn is less convenient?). So some basic tests of having torch/domain libraries/numpy operating okay (or not okay) is important, as people are going to try this.

Also, for this kinds of inference, being able to host the readonly model weights in memory (I guess important for both CPU and CUDA - for RAM my understanding is that it used to have shm, but I don't know if CUDA memory can be shared in some similar way) only once and have them accessible between multiple threads / subinterpreters may need to be discussed...

**Comment by albanD** [comment]
>  I don't know if CUDA memory can be shared in some similar way

Yes you can and that will be done by default if you send a cuda Tensor over multiprocessing objects.

> only once and have them accessible between multiple threads / subinterpreters

This is already possible for threads or processes based system today. Not sure what subinterpreter adds to the picture (beyond making PyObject ownership more complex since no object should be shared between interpreters)?

**Comment by vadimkantorov** [comment]
E.g. maybe some bugs like this? https://github.com/pytorch/pytorch/pull/113520

**Comment by albanD** [comment]
> maybe some bugs like this?

Nope, these registration are happening on the C++ lib which would be shared and would fail the same way.
I would argue this problem would actually be worse because you would loose the ability to lock between these concurrent calls since they can't interact with each other.

**Comment by vadimkantorov** [comment]
NumPy issue:
- https://github.com/numpy/numpy/issues/24755
